pkg/devfile/generator/utils.go

```diff
@@ -0,0 +1,170 @@
+
+		containerPorts = append(containerPorts, corev1.ContainerPort{
+			Name:          name,
+			ContainerPort: int32(endpoint.TargetPort),
+		})
+	}
+	return containerPorts
+}
+
+// getResourceReqs creates a kubernetes ResourceRequirements object based on resource requirements set in the devfile
+func getResourceReqs(comp v1.Component) corev1.ResourceRequirements {
+	reqs := corev1.ResourceRequirements{}
+	limits := make(corev1.ResourceList)
+	if comp.Container != nil && comp.Container.MemoryLimit != "" {
+		memoryLimit, err := resource.ParseQuantity(comp.Container.MemoryLimit)
+		if err == nil {
+			limits[corev1.ResourceMemory] = memoryLimit
+		}
+		reqs.Limits = limits
+	}
+	return reqs
+}
+
+// addSyncRootFolder adds the sync root folder to the container env and volume mounts
+func addSyncRootFolder(container *corev1.Container, sourceMapping string) string {
+	var syncRootFolder string
+	if sourceMapping != "" {
+		syncRootFolder = sourceMapping
+	} else {
+		syncRootFolder = DevfileSourceVolumeMount
+	}
+
+	// Note: PROJECTS_ROOT & PROJECT_SOURCE are validated at the devfile parser level
+	// Add PROJECTS_ROOT to the container
+	container.Env = append(container.Env,
+		corev1.EnvVar{
+			Name:  EnvProjectsRoot,
+			Value: syncRootFolder,
+		})
+
+	return syncRootFolder
+}
+
+// addSyncFolder adds the sync folder path for the container
+// sourceVolumePath: mount path of the empty dir volume to sync source code
+// projects: list of projects from devfile
+func addSyncFolder(container *corev1.Container, sourceVolumePath string, projects []v1.Project) error {
+	var syncFolder string
+
+	// if there are no projects in the devfile, source would be synced to $PROJECTS_ROOT
+	if len(projects) == 0 {
+		syncFolder = sourceVolumePath
+	} else {
+		// if there is one or more projects in the devfile, get the first project and check its clonepath
+		project := projects[0]
+		// If clonepath does not exist source would be synced to $PROJECTS_ROOT/projectName
+		syncFolder = filepath.ToSlash(filepath.Join(sourceVolumePath, project.Name))
+
+		if project.ClonePath != "" {
+			if strings.HasPrefix(project.ClonePath, "/") {
+				return fmt.Errorf("the clonePath %s in the devfile project %s must be a relative path", project.ClonePath, project.Name)
+			}
+			if strings.Contains(project.ClonePath, "..") {
+				return fmt.Errorf("the clonePath %s in the devfile project %s cannot escape the value defined by $PROJECTS_ROOT. Please avoid using \"..\" in clonePath", project.ClonePath, project.Name)
+			}
+			// If clonepath exist source would be synced to $PROJECTS_ROOT/clonePath
+			syncFolder = filepath.ToSlash(filepath.Join(sourceVolumePath, project.ClonePath))
+		}
+	}
+
+	container.Env = append(container.Env,
+		corev1.EnvVar{
+			Name:  EnvProjectsSrc,
+			Value: syncFolder,
+		})
+
+	return nil
+}
+
+// ContainerParams is a struct that contains the required data to create a container object
+type ContainerParams struct {
```

Comment:
Can we make this a non exported struct? It seems to be only used inside this package.